File: src/Database/Tests/SparseMatrixPropertyTest.cpp
/**
 * @file
 * @copyright This code is licensed under the 3-clause BSD license.\n
 *            Copyright ETH Zurich, Department of Chemistry and Applied Biosciences, Reiher Group.\n
 *            See LICENSE.txt for details.
 */
#include <Database/Exceptions.h>
#include <Database/Manager.h>
#include <Database/Objects/Model.h>
#include <Database/Objects/SparseMatrixProperty.h>
#include <gmock/gmock.h>

using namespace testing;
namespace Scine {
namespace Database {
namespace Tests {
/**
 * @class SparseMatrixPropertyTest SparseMatrixPropertyTest.cpp
 * @brief Comprises tests for the class Scine::Database::SparseMatrixProperty.
 * @test
 */
class SparseMatrixPropertyTest : public Test {
 public:
  SparseMatrixPropertyTest() {
  }
  static Credentials credentials;
  static Manager db;

  static void SetUpTestCase() {
    db.setCredentials(credentials);
    db.connect();
    db.init();
  }
  static void TearDownTestCase() {
    db.wipe();
  }
};
Credentials SparseMatrixPropertyTest::credentials(TEST_MONGO_DB_IP, std::atoi(TEST_MONGO_DB_PORT),
                                                  "unittest_db_SparseMatrixPropertyTest");
Manager SparseMatrixPropertyTest::db;

TEST_F(SparseMatrixPropertyTest, Create1) {
  // Setup
  auto coll = db.getCollection("properties");
  Eigen::SparseMatrix<double> data(3, 4);
  data.insert(0, 0) = 42.0;
  data.insert(0, 2) = 44.4;
  data.insert(2, 0) = 40.2;
  data.insert(2, 3) = 77.7;
  ID s, c;
  SparseMatrixProperty test =
      SparseMatrixProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), data, s, c, coll);
  ASSERT_TRUE(test.hasId());

  // Check Fields
  auto model = test.getModel();
  auto comment = test.getComment();
  auto name = test.getPropertyName();
  auto calculation = test.getCalculation();
  auto structure = test.getStructure();
  auto data_db = test.getData();

  ASSERT_EQ(model.method, "pbe");
  ASSERT_EQ(model.basisSet, "def2-svp");
  ASSERT_EQ(model.spinMode, "any");
  ASSERT_EQ(comment, "");
  ASSERT_EQ(name, "density_matrix");
  ASSERT_EQ(calculation, c);
  ASSERT_EQ(structure, s);
  ASSERT_EQ(data_db.coeff(0, 0), 42.0);
  ASSERT_EQ(data_db.coeff(0, 2), 44.4);
  ASSERT_EQ(data_db.coeff(2, 0), 40.2);
  ASSERT_EQ(data_db.coeff(2, 3), 77.7);
}

TEST_F(SparseMatrixPropertyTest, Create2) {
  // Setup
  auto coll = db.getCollection("properties");
  Eigen::SparseMatrix<double> data(3, 4);
  data.insert(0, 0) = 42.0;
  data.insert(2, 0) = 44.4;
  data.insert(0, 2) = 40.2;
  data.insert(2, 3) = 77.7;
  SparseMatrixProperty test =
      SparseMatrixProperty::create("density_matrix", Model("dft", "pbe", "def2-svp", "restricted"), data, coll);
  ASSERT_TRUE(test.hasId());

  // Check Fields
  auto model = test.getModel();
  auto comment = test.getComment();
  auto name = test.getPropertyName();
  auto data_db = test.getData();

  ASSERT_EQ(model.method, "pbe");
  ASSERT_EQ(model.basisSet, "def2-svp");
  ASSERT_EQ(model.spinMode, "restricted");
  ASSERT_EQ(comment, "");
  ASSERT_EQ(name, "density_matrix");
  ASSERT_FALSE(test.hasCalculation());
  ASSERT_FALSE(test.hasStructure());
  ASSERT_EQ(data_db.coeff(0, 0), 42.0);
  ASSERT_EQ(data_db.coeff(2, 0), 44.4);
  ASSERT_EQ(data_db.coeff(0, 2), 40.2);
  ASSERT_EQ(data_db.coeff(2, 3), 77.7);
}

TEST_F(SparseMatrixPropertyTest, Data) {
  // Setup
  auto coll = db.getCollection("properties");
  Eigen::SparseMatrix<double> data(3, 4);
  SparseMatrixProperty test = SparseMatrixProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), data, coll);
  ASSERT_TRUE(test.hasId());

  auto data_db = test.getData();
  ASSERT_EQ(data_db.coeff(0, 0), 0.0);
  ASSERT_EQ(data_db.coeff(2, 0), 0.0);
  ASSERT_EQ(data_db.coeff(0, 2), 0.0);
  ASSERT_EQ(data_db.coeff(2, 3), 0.0);
  data.insert(0, 0) = 42.0;
  data.insert(0, 2) = 44.4;
  data.insert(2, 0) = 40.2;
  data.insert(2, 3) = 77.7;
  test.setData(data);
  data_db = test.getData();
  ASSERT_EQ(data_db.coeff(0, 0), 42.0);
  ASSERT_EQ(data_db.coeff(0, 2), 44.4);
  ASSERT_EQ(data_db.coeff(2, 0), 40.2);
  ASSERT_EQ(data_db.coeff(2, 3), 77.7);
}

TEST_F(SparseMatrixPropertyTest, DataFails) {
  // Setup
  auto coll = db.getCollection("properties");
  SparseMatrixProperty test;
  Eigen::SparseMatrix<double> data;

  ASSERT_THROW(test.setData(data), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(test.getData(), Exceptions::MissingLinkedCollectionException);
  test.link(coll);
  ASSERT_THROW(test.setData(data), Exceptions::MissingIDException);
  ASSERT_THROW(test.getData(), Exceptions::MissingIDException);
}

} /* namespace Tests */
} /* namespace Database */
} /* namespace Scine */

File: src/Database/Python/Tests/sparse_matrix_property_test.py
# -*- coding: utf-8 -*-
__copyright__ = """This code is licensed under the 3-clause BSD license.
Copyright ETH Zurich, Department of Chemistry and Applied Biosciences, Reiher Group.
See LICENSE.txt for details.
"""

import scine_database as db
import unittest
import numpy as np
from scipy.sparse import rand

import os


class SparseMatrixPropertyTest(unittest.TestCase):
    def setUp(self):
        self.manager = db.Manager()
        self.manager.credentials.hostname = os.environ.get(
            'TEST_MONGO_DB_IP', '127.0.0.1')
        self.manager.credentials.port = int(os.environ.get('TEST_MONGO_DB_PORT', 27017))
        self.manager.credentials.database_name = "unittest_db_SparseMatrixPropertyTest"
        self.manager.connect()
        self.manager.init()

    def tearDown(self):
        self.manager.wipe()

    def test_make_one(self):
        # Setup
        coll = self.manager.get_collection("properties")
        c = db.ID()
        s = db.ID()
        model = db.Model("dft", "pbe", "def2-svp")
        ref = rand(1000, 10, density=0.2, format='csr')
        test = db.SparseMatrixProperty.make("density_matrix", model, ref, s, c, coll)
        assert test.has_id()

        # Check Fields
        model = test.get_model()
        comment = test.get_comment()
        name = test.get_property_name()
        calculation = test.get_calculation()
        structure = test.get_structure()
        data_db = test.get_data()

        assert model.method == "pbe"
        assert model.basis_set == "def2-svp"
        assert model.spin_mode == "any"
        assert comment == ""
        assert name == "density_matrix"
        assert calculation.string() == c.string()
        assert structure.string() == s.string()
        assert np.array_equal(data_db.todense(), ref.todense())

    def test_make_two(self):
        # Setup
        coll = self.manager.get_collection("properties")
        model = db.Model("dft", "pbe", "def2-svp")
        ref = rand(1000, 10, density=0.2, format='csr')
        test = db.SparseMatrixProperty.make("density_matrix", model, ref, coll)
        assert test.has_id()

        # Check Fields
        model = test.get_model()
        comment = test.get_comment()
        name = test.get_property_name()
        data_db = test.get_data()

        assert model.method == "pbe"
        assert model.basis_set == "def2-svp"
        assert model.spin_mode == "any"
        assert comment == ""
        assert name == "density_matrix"
        assert not test.has_calculation()
        assert not test.has_structure()
        assert np.array_equal(data_db.todense(), ref.todense())

    def test_create_one(self):
        # Setup
        coll = self.manager.get_collection("properties")
        c = db.ID()
        s = db.ID()
        model = db.Model("dft", "pbe", "def2-svp")
        test = db.SparseMatrixProperty()
        test.link(coll)
        ref = rand(1000, 10, density=0.2, format='csr')
        test.create(model, "density_matrix", s, c, ref)
        assert test.has_id()

        # Check Fields
        model = test.get_model()
        comment = test.get_comment()
        name = test.get_property_name()
        calculation = test.get_calculation()
        structure = test.get_structure()
        data_db = test.get_data()

        assert model.method == "pbe"
        assert model.basis_set == "def2-svp"
        assert model.spin_mode == "any"
        assert comment == ""
        assert name == "density_matrix"
        assert calculation.string() == c.string()
        assert structure.string() == s.string()
        assert np.array_equal(data_db.todense(), ref.todense())

    def test_create_two(self):
        # Setup
        coll = self.manager.get_collection("properties")
        model = db.Model("dft", "pbe", "def2-svp")
        test = db.SparseMatrixProperty()
        test.link(coll)
        ref = rand(1000, 10, density=0.2, format='csr')
        test.create(model, "density_matrix", ref)
        assert test.has_id()

        # Check Fields
        model = test.get_model()
        comment = test.get_comment()
        name = test.get_property_name()
        data_db = test.get_data()

        assert model.method == "pbe"
        assert model.basis_set == "def2-svp"
        assert model.spin_mode == "any"
        assert comment == ""
        assert name == "density_matrix"
        assert not test.has_calculation()
        assert not test.has_structure()
        assert np.array_equal(data_db.todense(), ref.todense())

    def test_data(self):
        # Setup
        coll = self.manager.get_collection("properties")
        s = db.ID()
        c = db.ID()
        model = db.Model("dft", "pbe", "def2-svp")
        ref = rand(1000, 10, density=0.2, format='csr')
        test = db.SparseMatrixProperty.make("density_matrix", model, ref, s, c, coll)
        assert test.has_id()

        assert np.array_equal(test.get_data().todense(), ref.todense())
        ref2 = rand(1000, 10, density=0.2, format='csr')
        test.set_data(ref2)
        assert np.array_equal(test.get_data().todense(), ref2.todense())

    def test_data_failure(self):
        # Setup
        coll = self.manager.get_collection("properties")
        test = db.SparseMatrixProperty()
        ref = rand(1000, 10, density=0.2, format='csr')
        self.assertRaises(RuntimeError, lambda: test.set_data(ref))
        self.assertRaises(RuntimeError, lambda: test.get_data())
        test.link(coll)
        self.assertRaises(RuntimeError, lambda: test.set_data(ref))
        self.assertRaises(RuntimeError, lambda: test.get_data())
